### PR TITLE
Initial support for command-line completion

### PR DESCRIPTION
**Pull request by LemonBoy**
Here's a working protoype that seems to work just fine with my synthetic test case.
In order to do as much as possible using Nim (that is, without creating any intermediate shell script) the completion protocol defined by [Bash](https://www.gnu.org/software/bash/manual/html_node/Programmable-Completion.html) is employed, it doesn't allow us to do fancy completion tricks but allows us to support Bash, Zsh and possibly any other compatible shell without adding more code/complexity.

The `ShellLexer` is needed to "parse" the input string and it does pass all the test-cases (courtesy of CPython `shlex` module).

## What's missing

- [ ] Test this harness with a real-world application
- [x] Option argument completions
  - I plan on adding another `pragma` to allow the user to specify a custom set of completions
- [x] Shell-compatible quoting/escaping routines
- [x] Completions for file/directories

## How to test this

```sh
# Only needed for zsh
autoload -Uz bashcompinit && bashcompinit
complete -C <binary-path> <command-name>
```

CC @zah

### CODE REVIEW

**Comment by zah** [comment]
How do you plan to handle argument-specific completions? In other words, if a parameter is accepting a file path or an enum value and the user has reached the point where these should be entered, the completions should be provided by a custom proc that handles the specific parameter type.

It seems to me that the main loop responsible for locating the currently active command and for invoking the correct setters needs only a small modification to handle completions as well:
https://github.com/status-im/nim-confutils/blob/ea8701d6be6c12b3224ea7c1380c6c3b3cba87a9/confutils.nim#L511-L550

The logic for traversing the command tree is the same and you just need to handle the situations where the command or option name is not yet fully entered. Just like we have custom setters, we'll have custom completion procs for handling the different param types.

**Comment by zah** [comment]
Otherwise, I like the command-line lexer and I should clarify that the user of `getopt()` is the main loop is just a temporary solution. Our goal is for Confutils to support the most command and native syntax for each platform where it runs.

**Comment by LemonBoy** [comment]
> How do you plan to handle argument-specific completions?

I think there are two different approaches:
- Having a `type`-attached `getArgCompletions` a-la `parseCmdArg`
- Having another pragma to specify a completion callback

The former maybe fits better the current confutils architecture since `parseCmdArg` /`default`/`hasDefault` are already using this pattern. The latter gives slightly more freedom since you don't couple the callback to the argument type (but sure, you can just use a `distinct T`).

> It seems to me that the main loop responsible for locating the currently active command and for invoking the correct setters needs only a small modification to handle completions as well:

The lexer is slightly more flexible and lenient, throwing the partial command line into `getopt` may or may not work as well.

> I should clarify that the use of getopt() is the main loop is just a temporary solution

There's not really anything that's intrinsically linked to how `getopt` parses the command line, I guess.

**Comment by zah** [comment]
I theory, we can have both the type-bound approach and the pragma-based overrides, but I would start with just the type-bound procs. They would allow a single developer who is more familiar with Confutils to create a library of types for a certain problem domain that the rest of the team will be able to reuse in multiple CLI tools while having only a basic familiarity with Confutils.

Otherwise, my main point is that you don't need two separate loops for processing the command-line in the two different scenarios (normal parsing and providing completions). Do you agree with that point? Perhaps I haven't thought thought the problem enough.

My suggestion is that we can replace `getopt` with a similar iterator based on the new lexer that will be able to handle the more traditional way of specifying options for Unix programs: `someutil -x foo -y bar`. And I've imagined that the Lexer will be parametric enough to be able to handle other styles as well (e.g. on Windows, it's typical for the arguments to be written like this: `someutil /x foo /y bar`)

**Comment by LemonBoy** [comment]
> Do you agree with that point?

I've tried that approach but had to backpedal because the code became too verbose whenever there's a difference in what to do due to the mode we're in (eg. A misspelled command name should not throw an error when trying to complete the cmdline).

> My suggestion is that we can replace getopt with a similar iterator based on the new lexer

That's orthogonal to this PR IMO, another lexer is possibly needed due to the many [differences](https://ss64.com/nt/syntax-esc.html) beside the obvious hypen-slash substitution.

File: confutils.nim [hunk 1]
```diff
@@ -220,6 +242,61 @@ proc parseCmdArgAux(T: type, s: TaintedString): T = # {.raises: [ValueError].} =
...
+      let match = if kind in {pcDir, pcLinkToDir}:
+        # Add a trailing slash so that completions can be chained
+        dir_path & DirSep & path & DirSep
+      else:
+        dir_path & DirSep & path
+
+      result.add(shellPathEscape(match))
+
+proc completeCmdArg*(T: type[OutFile|OutDir|OutPath], val: TaintedString): seq[string] =
+  return @[]
```
Comment (zah) [inline confutils.nim]: It's still possible to complete existing directory names here

File: confutils.nim [hunk 2]
```diff
@@ -422,6 +508,95 @@ proc load*(Configuration: type,
...
+  if len(completion) != 0:
+    var cmdStack = @[rootCmd]
+    # Try to understand what the active chain of commands is without parsing the
+    # whole command line
+    for tok in completion[1..^1]:
+      if not tok.startsWith('-'):
+        let subCmd = findSubcommand(cmdStack[^1], tok)
+        if subCmd != nil: cmdStack.add(subCmd)
+
+    let cur_word = completion[^1]
```
Comment (zah) [inline confutils.nim]: What about the cursor position btw? Is this reported by Bash somehow? I may be trying to complete something near the start of the line.

Comment (LemonBoy) [reply confutils.nim]: `COMP_POINT` reports the cursor position, we cut the input line (`COMP_LINE`) at that point and then parse it.

File: confutils.nim (hunk 1 above)
Comment (LemonBoy) [reply confutils.nim]: Makes sense, will add it.

File: confutils.nim (hunk 2 above)
Comment (zah) [reply confutils.nim]: Ah, I've missed that detail, thanks.

**Comment by zah** [comment]
For the completion of this feature, a reward of 420 USD has been awarded to LemonBoy:
https://etherscan.io/tx/0x343ae31ed4aec0944d5c6f5462dca63055d72e9ca96429e45ffe7cd56e307a9f

**Pull request merged**